rest/changes_api_test.go

```diff
@@ -1112,6 +1111,159 @@ func TestChangesActiveOnlyWithLimit(t *testing.T) {
 	response = it.SendAdminRequest("PUT", "/db/activeDoc0", `{"channel":["PBS"]}`)
 	assertStatus(t, response, 201)
 
+	response = it.SendAdminRequest("PUT", "/db/partialRemovalDoc", `{"channel":["PBS","ABC"]}`)
+	json.Unmarshal(response.Body.Bytes(), &body)
+	partialRemovalRev := body["rev"].(string)
+	assertStatus(t, response, 201)
+
+	response = it.SendAdminRequest("PUT", "/db/conflictedDoc", `{"channel":["PBS"]}`)
+	assertStatus(t, response, 201)
+
+	// Create a conflict, then tombstone it
+	response = it.SendAdminRequest("POST", "/db/_bulk_docs", `{"docs":[{"_id":"conflictedDoc","channel":["PBS"], "_rev":"1-conflictTombstone"}], "new_edits":false}`)
+	assertStatus(t, response, 201)
+	response = it.SendAdminRequest("DELETE", "/db/conflictedDoc?rev=1-conflictTombstone", "")
+	assertStatus(t, response, 200)
+
+	// Create a conflict, and don't tombstone it
+	response = it.SendAdminRequest("POST", "/db/_bulk_docs", `{"docs":[{"_id":"conflictedDoc","channel":["PBS"], "_rev":"1-conflictActive"}], "new_edits":false}`)
+	assertStatus(t, response, 201)
+
+	var changes struct {
+		Results  []db.ChangeEntry
+		Last_Seq interface{}
+	}
+
+	// Pre-delete changes
+	changesJSON := `{"style":"all_docs"}`
+	changesResponse := it.Send(requestByUser("POST", "/db/_changes", changesJSON, "bernard"))
+	err = json.Unmarshal(changesResponse.Body.Bytes(), &changes)
+	assertNoError(t, err, "Error unmarshalling changes response")
+	assert.Equals(t, len(changes.Results), 5)
```

Comment:
Is there potential races between calling `it.SendAdminRequest("POST", "/db/_bulk_docs"` and the changes from `it.Send(requestByUser("POST", "/db/_changes", changesJSON, "bernard"))` having all the expected changes?  In particular, during integration tests.

If so, can this be re-worked to do a retry loop so that it's robust against races?  I've been running into this issue a lot as part of https://github.com/couchbase/sync_gateway/issues/2938

Reply:
In another branch, I added some code that does a retry loop waiting for changes:

https://github.com/couchbase/sync_gateway/commit/5c46825d7bc0904da7fd8616622b00bdcf9b4715

It's the restTester rather than the indexTester though